server/aws-lsp-codewhisperer/src/language-server/utilities/telemetryUtils.ts

```diff
@@ -0,0 +1,108 @@
+    if (initializeParams?.initializationOptions?.aws) {
+        const { clientInfo } = initializeParams?.initializationOptions?.aws
+        const { extension } = clientInfo || {}
+
+        if (extension) {
+            items.push(`${extension.name ? format(extension.name) : 'UNKNOWN'}/${extension.version || 'UNKNOWN'}`)
+        }
+
+        if (clientInfo) {
+            items.push(`${clientInfo.name ? format(clientInfo.name) : 'UNKNOWN'}/${clientInfo.version || 'UNKNOWN'}`)
+        }
+
+        if (clientInfo?.clientId) {
+            items.push(`ClientId/${clientInfo?.clientId}`)
+        }
+    } else {
+        // Default to standard InitializeParams.clientInfo if no custom aws.clientInfo is set
+        const { clientInfo } = initializeParams || {}
+        if (clientInfo) {
+            items.push(`${clientInfo.name ? format(clientInfo.name) : 'UNKNOWN'}/${clientInfo.version || 'UNKNOWN'}`)
+        }
+    }
+
+    return items.join(' ')
+}
+
+const IDE_CATEGORY_MAP: { [key: string]: IdeCategory } = {
+    'AmazonQ-For-VSCode': 'VSCODE',
+    'Amazon-Q-For-JetBrains': 'JETBRAINS',
+    'AmazonQ-For-Eclipse': 'ECLIPSE',
+    'AWS-Toolkit-For-VisualStudio': 'VISUAL_STUDIO',
+}
+
+const mapClientNameToIdeCategory = (clientName: string): string | undefined => {
+    return IDE_CATEGORY_MAP[clientName]
+}
+
+// Use InitializeParams.initializationOptions.aws.clientInfo.extension to derive IDE Category from calling client
+// https://github.com/aws/language-server-runtimes/blob/main/runtimes/protocol/lsp.ts#L60-L69
+const getIdeCategory = (initializeParams: InitializeParams) => {
+    let ideCategory
+    if (initializeParams.initializationOptions?.aws.clientInfo?.extension?.name) {
+        ideCategory = mapClientNameToIdeCategory(initializeParams.initializationOptions.aws.clientInfo.extension.name)
+    }
+
+    return ideCategory || 'UNKNOWN'
+}
+
+// Map result from https://nodejs.org/api/process.html#process_process_platform to expected Operating system
+const getOperatingSystem = () => {
+    // Node.js environment
+    if (typeof process !== 'undefined') {
+        switch (process.platform) {
+            case 'darwin':
+                return 'MAC'
+            case 'win32':
+                return 'WINDOWS'
+            case 'linux':
+                return 'LINUX'
+            default:
+                return 'UNKNOWN'
+        }
+    }
+
+    return 'UNKNOWN'
+}
+
+// Compute UserContext object for sendTelemetryEvent API call.
+// Do not return context when unknown IDE or Operating system is found.
+// This behaviour may change in the future, when API will accept not enumerated values in API definition.
+export const makeUserContextObject = (initializeParams: InitializeParams, product: string): UserContext | undefined => {
+    const userContext: UserContext = {
+        ideCategory: getIdeCategory(initializeParams),
+        operatingSystem: getOperatingSystem(),
+        product: product,
```

Comment:
Which field should be populated by client/destination for the product field? and what can be the possible values of product?

Reply:
or should it be like from codeWhispererServer, we pass INLINE but for QServer, we pass CHAT?

Reply:
`product` will be set by every Server, which use TelemetryService, and it means specific Amazon Q Feature emitting telemetry (e.g. INLINE, CHAT).

Rest of fields are computed based on data set by Client following protocol.